core/src/main/java/org/apache/iceberg/io/CachingFileIO.java

```diff
@@ -0,0 +1,363 @@
+   */
+  private CachingFileIO() {
+  }
+
+  public CachingFileIO(FileIO io) {
+    assignIO(io);
+  }
+
+  public void assignIO(FileIO io) {
+    this.wrappedIO = io;
+  }
+
+  /**
+   * Initialize File IO from catalog properties.
+   * <p>
+   * It assumes that the inner FileIO is already initialized and therefore does not reinitialize it.
+   * If this is a reinitialization, it is the caller responsibility to call {@link #close()} and
+   * {@link #assignIO(FileIO)} prior to reinitialization to close the previously assigned FileIO and reassign a new one.
+   * Otherwise, previously assigned FileIO will be kept in use.
+   * <p>
+   * Reinitialization will discard previous content cache unless shared cache is being used
+   * ({@link CatalogProperties#IO_CACHE_SHARED} stays true).
+   * @param newProperties catalog properties
+   */
+  @Override
+  public void initialize(Map<String, String> newProperties) {
+    this.properties = newProperties;
+    initCache();
+  }
+
+  private void initCache() {
+    boolean isSharedCache = PropertyUtil.propertyAsBoolean(properties,
+        CatalogProperties.IO_CACHE_SHARED,
+        CatalogProperties.IO_CACHE_SHARED_DEFAULT);
+    if (isSharedCache) {
+      this.fileContentCache = getOrCreate(properties);
+      LOG.info("CachingFileIO created with shared cache. {} ", this.fileContentCache);
+    } else {
+      this.fileContentCache = createCache(properties);
+      LOG.info("CachingFileIO created. {} ", this.fileContentCache);
+    }
+  }
+
+  @Override
+  public void close() {
+    if (wrappedIO != null) {
+      wrappedIO.close();
+      wrappedIO = null;
+    }
+  }
+
+  @Override
+  public InputFile newInputFile(String path) {
+    InputFile inFile = wrappedIO.newInputFile(path);
+    if (fileContentCache != null && path.contains("/metadata/")) {
+      // TODO: Currently we just believe that this is a metadata file if it has "metadata" dir on its path. But metadata
+      //  location can be set differently for different tables. A table might have different metadata path by
+      //  setting custom "write.metadata.path" TableProperties.
+      //  We might need to extend FileIO interface. Add method 'newMetadataInputFile' and let caller use it to
+      //  differentiate metadata files vs other files.
+      return new CachingInputFile(fileContentCache, inFile);
+    }
+    return inFile;
+  }
+
+  @Override
+  public OutputFile newOutputFile(String path) {
+    return wrappedIO.newOutputFile(path);
+  }
+
+  @Override
+  public void deleteFile(String path) {
+    wrappedIO.deleteFile(path);
+    if (fileContentCache != null) {
+      fileContentCache.cache.invalidate(path);
+    }
+  }
+
+  public ContentCache getCache() {
+    return fileContentCache;
+  }
+
+  public long cacheEstimatedSize() {
+    return fileContentCache == null ? 0 : fileContentCache.cache.estimatedSize();
+  }
+
+  public static class CacheEntry {
+    private final int length;
+    private final List<ByteBuffer> buffer;
```

Comment:
I think this should be `buffers` since it is a list.

Also, if this class is private, then it's okay to access fields directly. But if it is public then everything should use getter methods. I don't see a reason for the class to be public, right?

Reply:
Yes, `CacheEntry` can be private.

I just wish that `ContentCache` itself is accessible to the outside world so that query engine can force cache invalidation by themself if necessary (ie., upon [REFRESH](https://impala.apache.org/docs/build3x/html/topics/impala_refresh.html) query).

Reply:
Done